include/picongpu/particles/manipulators/unary/FreeTotalCellOffsetRng.hpp

```diff
@@ -0,0 +1,147 @@
+                        }
+
+                    private:
+                        DataSpace<simDim> const m_superCellToLocalOriginCellOffset;
+                        RngType m_rng;
+                    };
+                } // namespace acc
+
+                template<typename T_Functor, typename T_Distribution>
+                struct FreeTotalCellOffsetRng
+                    : protected functor::User<T_Functor>
+                    , private functor::misc::TotalCellOffset
+                    , private functor::misc::Rng<T_Distribution>
+                {
+                    using CellOffsetFunctor = functor::misc::TotalCellOffset;
+                    using Functor = functor::User<T_Functor>;
+
+                    using RngGenerator = functor::misc::Rng<T_Distribution>;
+                    using RngType = typename RngGenerator::RandomGen;
+                    using Distribution = T_Distribution;
+
+                    template<typename T_SpeciesType>
+                    struct apply
+                    {
+                        using type = FreeTotalCellOffsetRng;
+                    };
+
+                    /** constructor
+                     *
+                     * @param currentStep current simulation time step
+                     */
+                    HINLINE FreeTotalCellOffsetRng(uint32_t currentStep)
+                        : Functor(currentStep)
+                        , CellOffsetFunctor(currentStep)
+                        , RngGenerator(currentStep)
+                    {
+                    }
+
+                    /** Create functor for the accelerator
+                     *
+                     * @tparam T_WorkerCfg lockstep::Worker, configuration of the worker
+                     * @tparam T_Acc alpaka accelerator type
+                     *
+                     * @param alpaka accelerator
+                     * @param localSupercellOffset offset (in superCells, without any guards) relative
+                     *                             to the origin of the local domain
+                     * @param workerCfg configuration of the worker
+                     */
+                    template<typename T_WorkerCfg, typename T_Acc>
+                    HDINLINE auto operator()(
+                        T_Acc const& acc,
+                        DataSpace<simDim> const& localSupercellOffset,
+                        T_WorkerCfg const& workerCfg) const -> acc::FreeTotalCellOffsetRng<Functor, RngType>
+                    {
+                        auto& cellOffsetFunctor = *static_cast<CellOffsetFunctor const*>(this);
+                        RngType const rng
+                            = (*static_cast<RngGenerator const*>(this))(acc, localSupercellOffset, workerCfg);
+                        return acc::FreeTotalCellOffsetRng<Functor, RngType>(
+                            *static_cast<Functor const*>(this),
+                            cellOffsetFunctor(acc, localSupercellOffset, workerCfg),
```

Comment:
Note: If you like to avoid casting you could write: `CellOffsetFunctor::operator()(acc, localSupercellOffset, workerCfg)`

Reply:
True, I copied from another wrapper and didn't think about it.